Sources/Swift/Metrics/BucketsMetricsAggregator.swift

```diff
@@ -0,0 +1,176 @@
+    private let dispatchQueue: SentryDispatchQueueWrapper
+    private let random: SentryRandomProtocol
+    private let totalMaxWeight: UInt
+    private let flushShift: TimeInterval
+    private let flushInterval: TimeInterval
+    private let flushTolerance: TimeInterval
+
+    private var timer: DispatchSourceTimer?
+    private var totalBucketsWeight: UInt = 0
+    private var buckets: [BucketTimestamp: [String: Metric]] = [:]
+    private let lock = NSLock()
+
+    convenience init(client: SentryMetricsClient, currentDate: SentryCurrentDateProvider, dispatchQueue: SentryDispatchQueueWrapper, random: SentryRandomProtocol) {
+
+        self.init(
+            client: client,
+            currentDate: currentDate,
+            dispatchQueue: dispatchQueue,
+            random: random,
+            totalMaxWeight: METRICS_AGGREGATOR_TOTAL_MAX_WEIGHT,
+            flushInterval: METRICS_AGGREGATOR_FLUSH_INTERVAL,
+            flushTolerance: METRICS_AGGREGATOR_FLUSH_TOLERANCE
+        )
+    }
+
+    init(
+        client: SentryMetricsClient,
+        currentDate: SentryCurrentDateProvider,
+        dispatchQueue: SentryDispatchQueueWrapper,
+        random: SentryRandomProtocol,
+        totalMaxWeight: UInt,
+        flushInterval: TimeInterval,
+        flushTolerance: TimeInterval
+    ) {
+        self.client = client
+        self.currentDate = currentDate
+        self.dispatchQueue = dispatchQueue
+        self.random = random
+
+        // The aggregator shifts its flushing by up to an entire rollup window to
+        // avoid multiple clients trampling on end of a 10 second window as all the
+        // buckets are anchored to multiples of ROLLUP seconds. We randomize this
+        // number once per aggregator boot to achieve some level of offsetting
+        // across a fleet of deployed SDKs.
+        let flushShift = random.nextNumber() * ROLLUP_IN_SECONDS
+        self.totalMaxWeight = totalMaxWeight
+        self.flushInterval = flushInterval
+        self.flushShift = flushShift
+        self.flushTolerance = flushTolerance
+
+        startTimer()
+    }
+
+    private func startTimer() {
+        let timer = DispatchSource.makeTimerSource(flags: [], queue: dispatchQueue.queue)
+
+        // Set leeway to reduce energy impact
+        let leewayInMilliseconds: Int = Int(flushTolerance * 1_000)
+        timer.schedule(deadline: .now() + flushInterval, repeating: self.flushInterval, leeway: .milliseconds(leewayInMilliseconds))
+        timer.setEventHandler { [weak self] in
+            self?.flush(force: false)
+        }
+        timer.activate()
+        self.timer = timer
+    }
+
+    func add(type: MetricType, key: String, value: Double, unit: MeasurementUnit, tags: [String: String]) {
+
+        // It's important to sort the tags in order to
+        // obtain the same bucket key.
+        let tagsKey = tags.sorted(by: { $0.key < $1.key }).description
+        let bucketKey = "\(type)_\(key)_\(unit.unit)_\(tagsKey)"
```

Comment:
h:
For tags as `["some": "tag", "and": "another-one"]` `tagsKey` will be `[(key: "and", value: "another-one"), (key: "some", value: "tag")]`

This means `bucketKey` also depends on the tags values. is this right?

It this is by design can we format the `tagsKey` to a more human readable value?

```suggestion
        let tagsKey = tags.sorted(by: { $0.key < $1.key }).map( { "\($0.key)=\($0.value)" } ).joined(separator: ",") 
        let bucketKey = "\(type)_\(key)_\(unit.unit)_\(tagsKey)"
```

Reply:
The only requirement for the bucketKey is that it has to be unique for type, key, unit, and tags. 
```swift
// current version
"counter_key1_day_[(key: \"and\", value: \"another-one\"), (key: \"some\", value: \"tag\")]"

// your version
"counter_key1_day_and=another-one,some=tag"
```
You version looks nicer, so I will update it thanks 😃